### ISSUE TITLE
Consider using FSLEYES plugin as wrapper for the Shimming Toolbox

### ISSUE THREAD

**Issue by po09i** [shimming-toolbox/shimming-toolbox#205]
### Feature Description:
Shimming Toolbox will eventually have a GUI. We could leverage FSLEYES to display the functionalities of the toolbox and make the CLIs available from the FSLEYES' GUI. Here is a screenshot from SCT.

![Screen Shot 2021-01-12 at 1 49 59 PM](https://user-images.githubusercontent.com/47249340/104411073-c1ec8e80-5537-11eb-85bc-a6f3617b1782.png)

Pros:
- Easier implementation than most options
- Already deals with niftis
- Also in python

Cons:
- Requires the end user to have fsleyes installed to use the GUI
- Hidden feature in fsleyes (maybe)

Some considerations if we go ahead with FSLeyes:
- have a design (put it in a platform where it's easy to give feedback/iterate-- draw.io, gslide, etc) before starting implementing
- could be either a script (like in [SCT](https://github.com/neuropoly/spinalcordtoolbox/blob/master/contrib/fsl_integration/sct_plugin.py)) or a plugin (like in [ADS](https://github.com/neuropoly/axondeepseg/blob/master/ads_plugin.py))
- installation: should we add fsleyes as a dependency in the requirements?
- license compat: https://git.fmrib.ox.ac.uk/fsl/fsleyes/fsleyes/-/blob/master/LICENSE

Feel free to add your pros and cons

**Comment by charjl** [comment]
-ADS has a plugin for this (FSL-Eyes), SCT has a script. Why the difference?
-Is FSL-Eyes the correct tool for this ? There are installation problems. <2020-01-26: https://docs.google.com/document/d/1CKpGLPcQ133hNdVa-Hzk5AINkJiQ13-uBEiDogpE8sg/edit#heading=h.cuol6iqlzhv8 > 
-Different cli have tabs and images on the SCT one
-Implement it in such a way where you don't have to load the script into fsleyes, that it starts loaded from command line
-Run FSLEyes in SCT to get used to it
-Look at ADS implementation. Take what we want from each of the implementations

**Comment by po09i** [comment]
The main CLIs we want to implement are:
- st_mask
- st_dicom_to_nifti
- st_prepare_fieldmap
- st_realtime_zshim

**Comment by po09i** [comment]
### How should the main layout be? 
ADS and SCT's implementation is included for reference and to show the different tabs:

![Screen Shot 2021-02-18 at 16 34 38](https://user-images.githubusercontent.com/47249340/108424447-2ec60900-7207-11eb-8e19-adb1f277283c.png)
![Screen Shot 2021-02-18 at 16 35 09](https://user-images.githubusercontent.com/47249340/108424495-41d8d900-7207-11eb-9282-471b0e1522ed.png)

### Current options
- One big tab: like ADS and Jason’s Matlab implementation
  - This would mean different portions of the tab would be allocated to the different main uses of the toolbox: shim, dcm2nifti, fieldmap, mask
  - Dropdowns would dynamically update the “buttons” available for the appropriate option selected
- 1 tab per CLI
  - This might become not user friendly in the future when we add more shimming experiments.
  - Divide in smaller, easy to digest features
- 4 tabs: shim, dcm2nifti, fieldmap, mask
  - Start on the shimming tab which would list the available experiments, update the tab once selected to display the proper requirements of the experiment.
  - If for example a fieldmap is required, then the user switches to the fieldmap tab and creates it so that they can continu on the shim tab. 
  - Each tab could update dynamically (mask would have a dropdown: box, spinal cord, brain etc..)
	
In a meeting, some members of the team seemed to like the 4 tabs option a bit more. A sketch will be done to ecplore futher. Feedback on the options is appreciated.

**Comment by jcohenadad** [comment]
my 2-cent:
- I do have a preference for the horizontal layout with tabs (à la SCT)
- 1 tab per CLI --> definitely not
- 4 tabs: shim, dcm2nifti, fieldmap, mask --> seems like a good start
- start simple prototype and we iterate from there

**Comment by po09i** [comment]
Here's a link to the initial sketch: https://drive.google.com/file/d/1dETgz9SbKZ5ktNwi9NE-Kh1LDuFayzcV/view?usp=sharing

**Comment by jcohenadad** [comment]
> Here's a link to the initial sketch: https://drive.google.com/file/d/1dETgz9SbKZ5ktNwi9NE-Kh1LDuFayzcV/view?usp=sharing

nice! comments added

**Comment by charjl** [comment]
If one uses a fsleyes plugin, fsleyes must be run from inside our shimming conda virtual environment (often called shim_venv)
When fsleyes runs from inside this environment, dependencies for shimming-toolbox are overwritten, and shimming pytest cannot be run. 
The other option, which is quite similar to the eye, is to make a fsleyes script rather than a fsleyes plugin. Fsleyes scripts can be run within an flseyes that is not in our virtual environment. I have verified that no dependencies are overwritten in the case of loading a script.

**Comment by jcohenadad** [comment]
script is what we use for SCT-- i don't see any problem with that approach (apart from the fact that we need to load it each time-- which literally takes 2s)

**Comment by ahill187** [comment]
So Paul McCarthy replied and says he's releasing a new version of `fsleyes` in a couple of weeks and will include our `dcm2niix` dependency fix:
https://github.com/pauldmccarthy/fsleyes/issues/69

**Comment by jcohenadad** [comment]
awesome news! thank you for checking @ahill187

**Comment by charjl** [comment]
- [x] Finder/file browser (Ainsleigh)
- [x] "Run" implementing on all four tabs (Char/Alex)
- [x] Determine what other input types are needed mostly strings, some paths, numbers (Gaspard)
- [x] Second type-specific file finder from overlay list (Alex)
- [x] Basic call-back infrastructure for drop-down choice Drop-down list for types of shimming (Alex/Gaspard)
- [x] Third column for terminal output (Char/Ainsleigh)
- [x] Tab descriptions (Alex)
- [x] Add default paths for configpath variables (Ainsleigh/Alex)
- [x] Dropdown for shape types (Alex/Char)
- [x] Add dropdown for `unwrapper` - inputs will change if we select `prelude` (currently only avail.) (Alex)
- [x] Add dropdown for box/rect/threshold (similar to shimming) for `mask` tab (Ainsleigh)
- [x] Update echo input rows based on `number of echoes` selected
- [x] Add more verbosity to terminal (figure out how to get better errors from subprocess)
- [x] Create output folder if it doesn't exist (Ainsleigh)
- [x] Figure out how to do multiple arguments for the Mask tab item `center` (Ainsleigh)
- [x] Add help icons (?) to describe each field (Ainsleigh)
- [x]  Add `is_required` field in `input_text_box_metadata` for CLI params that have default values (Ainsleigh)
- [x] Add output files back in the overlay (Alex/Ainsleigh)
- [ ] Add tests
- [ ] Bug with dropdown, the icons (asterisk and info icons) don't reload properly
- [x] Add `is_required` field for phase once implemented

**Comment by charjl** [comment]
Fiesta has been completed; implementation to begin.

**Comment by gaspardcereza** [comment]
# Inputs used by each CLI
## dicom_to_nifti
- inputs:
        -input (path): path to the input dicom folder. **Required**
        -output (path): path to the output nifti folder. 
        -subject (str): Name of the patient. **Required**
        -config (path): Full file path and name of the bids config file

## mask 
- inputs:
      (str): 'box', 'rect' or 'threshold'. Type of masking to apply

#### mask --> 'box' & 'rect'
- additionnal inputs:
      -input (path): Input path of the nifti file to mask. This nifti file must have 3D. Supported extensions are .nii or .nii.gz. **Required**
      -output (path): Name of output mask. Supported extensions are .nii or .nii.gz. 
      -size (int, int, int): Length of the side of the box along first, second and third dimension (in pixels). **Required**
      -center (int, int, int): Center of the box along first, second and third dimension (in pixels).

#### mask --> 'threshold'
- additionnal inputs:
      -input (path): Input path of the nifti file to mask. This nifti file must have 3D. Supported extensions are .nii or .nii.gz.
      -output (path): Name of output mask. Supported extensions are .nii or .nii.gz.
      - thr (int): Value to threshold the data: voxels will be set to zero if their value is equal or less than this threshold.

## prepare Fieldmap
- inputs:
      phase (path): Input path of phase nifti file(s), in ascending order: echo1, echo2, etc. **Required**
      -mag (path): Input path of mag nifti file. 
      -unwrapper (str): Algorithm for unwrapping.
      -output (path): Output filename for the fieldmap, supported types : '.nii', '.nii.gz'
      -mask (path): Input path for a mask. Used for PRELUDE
      -threshold (float): Threshold for masking. Used for: PRELUDE
      -gaussian-filter (bool): Gaussian filter for B0 map
      -sigma (float): Standard deviation of gaussian filter. Used for: gaussian_filter

## shim --> realtime_zshim
- inputs:
      -fmap (path): B0 fieldmap. This should be a 4d file (4th dimension being time. **Required**
      -anat (path): Filename of the anatomical image to apply the correction. **Required**
      -resp (path): Siemens respiratory file containing pressure data. **Required**
      -mask-static (path): 3D nifti file with used to define the spatial region to shim. The coordinate system should be the same as ``anat``'s coordinate system.
      -mask-riro (path): 3D nifti file with used to define the spatial region to shim. The coordinate system should be the same as ``anat``'s coordinate system.
      -output (path):Directory to output gradient text file and figures.
      -verbose (bool?): Be more verbose.

@po09i would be great if you could have a quick look to this in order to double check if everything is ok.

**Comment by jcohenadad** [comment]
Why I prefer script over plugin?
- because the plugin gets "installed" and a file is created under `~/.fsleyes`
- related to the previous point: if ST is updated, the user has the responsibility to update the plugin. Bottomline: users won't do it.
- fsleyes needs to be installed inside the venv (big cons!)

A cons of the script is that we need to manually load it each time, however this is not a big deal as it could be included in the CLI call to fsleyes, and copy/pasted from our ST RTD.
